### DISCUSSION TITLE
Welcome to lithium Discussions!

### DISCUSSION THREAD

**Discussion by matt-42** [matt-42/lithium#45] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by cong1920** [comment]
As you are the project owner, what are the top 3 features you plan to add?
And what are the top 3 improvements in your plan?

Asking because I, as a trivial contributor, might just randomly feel something needs to be added or done better, and then start coding so. For example, I am trying to add default index file list for static serving because I just feel if Lithium is used to host a HTML5 SPA app it should be more user friendly to use https://web-app-site/ directly without typing index.html in the end. But this kind of random contributions might make Lithium off your planned track more and more.

**Comment by matt-42** [comment]
Hi @cong1920 , you are welcome to add features for static file serving. Its something lithium should be able to do.

There is no roadmap yet, up until now, it was to align the performance with the fastest webserver available (otherwise
there is no point of coding it in c++) and to finish the documentation website. Both are done.

My idea was to wait for users to ask for missing features (as you did :) so we can implement them. Lithium is pretty new so it
does not get much feedback yet, but hopefully by ranking #1 in the next techempower round20 we'll reach more users.

**Reply by cong1920** [reply]
Merry Christmas and Happy New Year!

My understanding of priority 1 by now is to get and keep ranking #1 in the TechEmpower benchmark tests. That is also where I heard about this project and gave it a try. Not sure when round 20 result will come out. But from round 19 result Lithium is behind because of DB operations. Hope you had made improvements in that area. Will definitely keep an eye on the benchmark result.

Besides documentations, do you think some tutorial like posts can be more helpful for new users to ramp up?

**Reply by matt-42** [reply]
Happy new year @cong1920 ! The next TFB round will be released very soon. Lithium should be #1. Some tutorial will definitely help to get more users yes. Do you have a blog or do you want to post the article on lithium's website ? (there is no lithium blog yet but I can add one).

**Reply by cong1920** [reply]
Hmm, I am planning to write some tutorial like posts, even I haven't tried any advanced features of Lithium yet. But it could be my learning track so not a waste of text I typed in :). Once I have some drafts ready I will send you to review first.

One more question which is about code. What is general rule of utilizing 3rd party libs? I am writing more tests but I feel like some mature test lib could be more helpful like the [Catch](https://github.com/catchorg/Catch2). It is header-only lib therefore it does not introduce in any binary dependency for building. Even more how about a mature logging lib other than std::cout? Like the [spdlog](https://github.com/gabime/spdlog), which is also a header-only lib.

**Reply by matt-42** [reply]
Hi @cong1920,

Thanks for the proposition. I'm looking forward to review your draft !

What features of Catch do you need for testing ? Or what problem did you have with the Ctest + assert approach ? I'm open to add a 3rd test lib only if you can point issues with the current test suite.
I'm saying this because Catch is amazingly slow to compile (10s with gcc for their simple factorial example https://github.com/catchorg/Catch2/blob/v2.x/docs/tutorial.md#top) and will slow down our productivity. 

Same for the log, nothing is logged for now except fatal errors and cout/cerr was fine. If you can point issues with this, please share them here so we can see how to solve them.

**Reply by cong1920** [reply]
Sorry for belated reply. I am kinda busy preparing interviews these days. Here is an outline of contents of the tutorial series that I planned.

1. Introduction to web app programming in C++ and Lithium
2. What readers should know already
3. Tools/Libs to get start
    3.1. Gcc, CMake, and Docker (Docker with WSL2 on Windows)
    3.2. Libboost-dev, libmysql, libpq and libopenssl
4. Build "hello, world!" with Lithium
5. Build a web server serving static contents
    5.1. Plain HTML page
    5.2. Parsing request parameters from query string
    5.3. Parsing form data from POST request
6. REST APIs
7. Work with database
8. Websocket
9. Examples
    9.1 Auth
    9.2 TODO workitem tracking web app
    9.3 Blog system
...
...

Of course there must be something missed or I skipped as I don't fully understand yet, e.g. async coding with fiber, symbol based programming paradigm, and etc. Please feel free to add what you believe must be included in tutorial.

**Reply by matt-42** [reply]
Thanks @cong1920 for the propositions. That's a lot of work :), here is my review below. I added  link to the official docs where I think this may be a duplicate. 

On the other hand, I like the idea of the blog system tutorial. The official docs are actually still missing complete examples, so you could focus on them. Here what I see as candidates for examples: A rest API and a blog system. For the auth system, I'm not sure if this is relevant if the blog system includes user authentication.

Just one question: Would you prefer adding these tutorials in the official documentation web site or in a personal blog ?

    [ Introduction to web app programming in C++ and Lithium
    What readers should know already
    Tools/Libs to get start 
    3.1. Gcc, CMake, and Docker (Docker with WSL2 on Windows) 
    3.2. Libboost-dev, libmysql, libpq and libopenssl  ]
         -> there is already a getting started section in the doc. https://matt-42.github.io/lithium/getting-started 
             Could you check it and see if there is something missing ?
        -> same remark, see the getting started.

    Build "hello, world!" with Lithium
          -> same remark, check this page: https://matt-42.github.io/lithium/http-server/introduction
    Build a web server serving static contents,     5.1. Plain HTML page
          -> there is this page already: https://matt-42.github.io/lithium/http-server/serving-static-files
    5.2. Parsing request parameters from query string
    5.3. Parsing form data from POST request
          -> https://matt-42.github.io/lithium/http-server/get-post-and-url-request-parameters
   
    REST APIs
         -> there is no example of rest API, so this may be a nice candidate for an example
    Work with database
         -> https://matt-42.github.io/lithium/sql/headers

    Websocket -> there is no websocket support yet.
    Examples
    9.1 Auth -> (already included in the blog system ?)
    9.2 TODO workitem tracking web app (what do you mean by tracking web app?)
    9.3 Blog system  -> good candidate for an example

**Reply by cong1920** [reply]
Feeling so sorry, even guilty, for my belated reply. Besides preparing resume and interviews, there is a home remodeling project going on. Still I should really open GitHub often and start looking into your links during my fragmentary time pieces.

First to answer the question of "adding these tutorials in the official documentation web site or in a personal blog ?". It's really up to you. I am open to both but I don't have my own blog yet :) It is my honor to contribute to official doc website if that works for you better.

My original thought about the tutorial is to make reader realize how easy and how similar with NodeJS and other framework of script programming when writing web application with Lithium in native C++. So some of my examples might seem too basic or kinda redundant like the separated "auth" and then "blogging". And also I didn't know there is no websocket support yet. I will remove that section. Thanks for pointing out there is no REST APIs example, probably I should start with that part to draft and send you to review first to see if my writing is okay :)

Oh, for "TODO workitem tracking web app", sorry for confusing. I just mean a simple web app that can let users track TODO workitems. It is as simple as just like for signed in users, it can create work items, change status as "in progress", and mark "completed". In my mind it is simpler than a blogging system, but it needs "auth" module built ready first. 

Hope soon I will be free from my tediously LeetCoding days and get more time for this fun project.

**Reply by matt-42** [reply]
No worries! I'm also busy on my current job and my time for lithium is also limited.
I like the idea of an article that help javascript users to get into lithium. 

I forgot to tell you but there is a blog system example, but just code + comments, not a blog post.
https://github.com/matt-42/lithium/blob/master/libraries/http_server/examples/blog.cc

I suggest that you focus on just one thing if your time is limited: one of the blog articles you proposed (from js to lithium, or the blog system), or moving the test suite to doctest.

If it's a blog post, it will be great to post it on the official website !

Cheers,
Matt

**Reply by cong1920** [reply]
Yes, sir. Focusing on one thing first. I think a good start point for me is that I can start composing a step-by-step tutorial post from one of your code + comments example.

**Reply by jmafc** [reply]
Hi @matt-42,

Regarding your earlier comment:

> I'm saying this because Catch is amazingly slow to compile (10s with gcc for their simple factorial example https://github.com/catchorg/Catch2/blob/v2.x/docs/tutorial.md#top) and will slow down our productivity.

I ran into this too, but then someone pointed out that I should create a separate `test_main.cpp` file with a simple #define and #include "catch2.hpp". The object from that file then gets linked with your other code/tests and doesn't add 10s to each compile/test cycle. However, after learning that I found `doctest` and I believe it's even faster than `catch2`. And you can use the same trick, i.e., have a test_main.cpp like this:
```c++
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>
```
`doctest` has even more macros if you want to really squeeze machine cycles from testing.

**Comment by cong1920** [comment]
And for using Catch in test code, and considering logging lib. Sorry I didn't realize compiling could be hugely slowed down with Catch. I like to use Catch in test because:
1) Catch can group test scenarios better. Like we can have good test scenarios in one group, and error cases in another.
2) Catch can get test setup/teardown better than assert. At least for the test code I added to clean up temporary files assert() on failing cases never have chance clean them up because assert() just terminates process directly.

If we stay away with Catch, could here be a way to have test groups in one test code file with ctest? and a gently teardown than assert()? Or any mitigation like build flag or something to only build with Catch for test code if it turned on?

Same for logging I do believe logging with timestamps and output redirection to file or screen is pretty important for a web server. Or maybe it should be Lithium user's choice to use a modern logging lib when coding a web server with Lithium?

**Comment by jmafc** [comment]
@matt-42  Thought you may want to know that, although the logo is clearly visible on https://matt-42.github.io/lithium/ no matter which dark/light theme you use, on GH itself, the logo in the README is nearly invisible if you use a dark theme.
